### DISCUSSION TITLE
sysex - MIDI hardware vs USB

### DISCUSSION THREAD

**Discussion by Lonolo** [probonopd/MiniDexed#633] [Q&A]
I can use DEXED or https://dx7.vstforx.de/ to change every operator's voice parameters and see it happen on the MiniDexed module, but it only works with hardware MIDI for me.

I can connect, send note and CC, but no sysex param changing over USB: Gadget or host. Was wondering if anyone else had success or is this not implemented for USB, just with hardware MIDI only?

One difference I see in the HDMI debug is the messages are delivered differently:

USB-Gadget output:
<img width="742" alt="Screen Shot 2024-04-01 at 8 41 07 AM" src="https://github.com/probonopd/MiniDexed/assets/1200412/8ba40aa2-5271-474e-8465-f417fa697d1f">

Hardware-MIDI output:
<img width="728" alt="Screen Shot 2024-04-01 at 8 43 03 AM" src="https://github.com/probonopd/MiniDexed/assets/1200412/e23f853a-528c-44c9-8a92-1d4a6654a64b">

**Comment by Banana71** [comment]
Hi @Lonolo,

Even with my miniDexed in gadget mode (Raspberry Pi Zero 2W) I cannot parameterize the patches using Dexed. 😞

**Comment by diyelectromusic** [comment]
Hmm.  The log is different as there is extra logging in the MIDI serial device code as well as in the message handler, but both serial and USB pass the data straight onto the message handler - e.g. https://github.com/probonopd/MiniDexed/blob/main/src/midikeyboard.cpp#L106

But now I think about it, I believe there is something slightly different about SysEx with USB as it won't all fit in a single USB MIDI packet, so it may be something to do with the USB MIDI driver in circle - maybe it needs additional piecing together or something.  I'll have to take a look.

Yes, seeing how the packet handler is called (https://github.com/rsta2/circle/blob/master/lib/usb/usbmidi.cpp#L241), it looks like a multi-USB packet (such as a SysEx) would involve multiple calls to the packet handler, so we might need additional logic in the USB handler to glue together a SysEx message before passing onto our message handler....  Hmm..

Kevin

**Reply by diyelectromusic** [reply]
Here is a possible solution, but it hasn't been tested very thoroughly by me so far.  Would be great if anyone is able to give it a go in various configurations.

https://github.com/probonopd/MiniDexed/pull/634

Kevin
